File: pset6/credit.py
def main():
    
    # obtain credit card number
    n = int(input("Number: "))
    
    # determine if the card number is valid
    validate(n)
    
    # determine the associated credit card company
    company(n)

# define validate function
def validate(n):
    
    sumcard = 0;
    
    i = 10
    while i <= n:
        one = ((n // i) % 10) * 2
        i *= 100
        
        j = 1
        while j <= one + 1:
            two = (one // j) % 10
            sumcard += two
            j *= 10
    
    k = 1
    while k <= n:
        three = ((n // k) % 10)
        sumcard += three
        k *= 100
    
    if ((int(sumcard) % 10) != 0):
        print("Invalid")
        
# define company function
def company(n):            
    # Amex
    if (n >= 340000000000000 and n < 1000000000000000):
        print("AMEX")
        
    # MasterCard
    elif (n >= 5100000000000000):
        print("MASTERCARD")
        
    # VISA
    elif (n >= 4000000000000 or n >= 4000000000000000):
        print("VISA")
        
# initiate main function
if __name__ == "__main__":
    main()

File: pset1/credit.c
#include<cs50.h>
#include<stdio.h>

void validate(long long n);
void company(long long n);

int main(void)
{
    printf("Number: ");
    long long n = get_long_long();
    
    validate(n);
    
    company(n);
}

void validate(long long n)
{
    long long sum = 0;
    
    for (long long  i = 10; i <= n; i *= 100)
    {
        long long one = (((n / i) % 10) * 2);
        
        for (long long j = 1; j <= one + 1; j *= 10)
        {
            long long two = (one / j) % 10;
            sum = sum + two;
        }
    }
    
    for (long long k = 1; k <= n; k *=100)
    {
        long long three = ((n / k) % 10);
        sum = sum + three;
    }
    
    if ((sum % 10) != 0)
    {
        printf("INVALID\n");
    }
}

void company(long long n)
{
    // AMEX
    if (n >= 340000000000000 & n < 1000000000000000)
    {
        printf("AMEX\n");
    }
    // MasterCard
    else if (n >= 5100000000000000)
    {
        printf("MASTERCARD\n");
    }
    // VISA
    else if (n >= 4000000000000 || n >= 4000000000000000)
    {
        printf("VISA\n");
    } 
}